File: frontend/src/api/paper.ts
import client from "./client";

export interface PaperPosition {
  id: number;
  symbol: string;
  qty: number;
  avg_cost: number;
  current_price: number;
  market_value: number;
  cost_basis: number;
  unrealized_pnl: number;
  unrealized_pnl_pct: number;
  day_pnl: number;
  day_pnl_pct: number;
  prev_close: number;
  opened_at: string;
}

export interface PaperAccount {
  cash: number;
  starting_balance: number;
  equity: number;
  day_pnl: number;
  day_pnl_pct: number;
  total_pnl: number;
  total_pnl_pct: number;
  created_at: string;
  positions: PaperPosition[];
}

export interface PaperOrder {
  id: number;
  symbol: string;
  side: "buy" | "sell";
  qty: number;
  notional: number | null;
  order_type: string;
  limit_price: number | null;
  stop_price: number | null;
  take_profit_price: number | null;
  stop_loss_price: number | null;
  tif: string;
  status: "working" | "filled" | "cancelled" | "rejected" | "expired";
  fill_price: number | null;
  fill_qty: number | null;
  slippage: number | null;
  total: number;
  reject_reason: string | null;
  filled_at: string | null;
  created_at: string;
}

export interface PlaceOrderBody {
  symbol: string;
  side: "buy" | "sell";
  qty?: number;
  notional?: number;
  order_type?: "market" | "limit" | "stop" | "stop_limit" | "bracket" | "oco" | "trailing_stop";
  limit_price?: number | null;
  stop_price?: number | null;
  take_profit_price?: number | null;
  stop_loss_price?: number | null;
  trailing_amount?: number | null;
  trailing_type?: "amount" | "percent" | null;
  tif?: "day" | "gtc" | "ioc" | "fok";
  extended_hours?: boolean;
  idempotency_key?: string;
}

export interface PaperTransaction {
  id: number;
  symbol: string;
  side: string;
  qty: number;
  fill_price: number;
  cost_basis: number;
  realized_pnl: number;
  created_at: string;
}

export interface DailySnapshot {
  date: string;
  equity: number;
  cash: number;
  positions_value: number;
  day_pnl: number;
}

export async function getAccount(): Promise<PaperAccount> {
  const { data } = await client.get<PaperAccount>("/paper/account");
  return data;
}

export async function placeOrder(body: PlaceOrderBody): Promise<PaperOrder> {
  const { data } = await client.post<PaperOrder>("/paper/orders", body);
  return data;
}

export async function getOrders(status?: string): Promise<PaperOrder[]> {
  const { data } = await client.get<PaperOrder[]>("/paper/orders", {
    params: status ? { status } : undefined,
  });
  return data;
}

export async function cancelOrder(id: number): Promise<{ ok: boolean }> {
  const { data } = await client.delete<{ ok: boolean }>(`/paper/orders/${id}`);
  return data;
}

export async function getTransactions(): Promise<PaperTransaction[]> {
  const { data } = await client.get<PaperTransaction[]>("/paper/transactions");
  return data;
}

export async function getSnapshots(): Promise<DailySnapshot[]> {
  const { data } = await client.get<DailySnapshot[]>("/paper/snapshots");
  return data;
}

export async function resetAccount(): Promise<{ ok: boolean }> {
  const { data } = await client.post<{ ok: boolean }>("/paper/reset");
  return data;
}

export async function seedDemo(): Promise<{ ok: boolean }> {
  const { data } = await client.post<{ ok: boolean }>("/paper/seed-demo");
  return data;
}

File: backend/app/db/models/paper.py
from __future__ import annotations
from datetime import datetime, timezone
from decimal import Decimal
from typing import Optional
from sqlalchemy import DateTime, Float, Integer, String, Boolean, JSON, func
from sqlalchemy.orm import Mapped, mapped_column
from app.db.base import Base

STARTING_BALANCE = 100_000.0

class PaperAccount(Base):
    __tablename__ = "paper_accounts_archived"
    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    user_id: Mapped[int] = mapped_column(Integer, nullable=False, index=True, unique=True)
    cash: Mapped[float] = mapped_column(Float, nullable=False, default=STARTING_BALANCE)
    starting_balance: Mapped[float] = mapped_column(Float, nullable=False, default=STARTING_BALANCE)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=lambda: datetime.now(timezone.utc), server_default=func.now())

class PaperPosition(Base):
    __tablename__ = "paper_positions_archived"
    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    user_id: Mapped[int] = mapped_column(Integer, nullable=False, index=True)
    symbol: Mapped[str] = mapped_column(String, nullable=False)
    qty: Mapped[float] = mapped_column(Float, nullable=False)
    avg_cost: Mapped[float] = mapped_column(Float, nullable=False)
    prev_close: Mapped[Optional[float]] = mapped_column(Float, nullable=True)  # for day P&L
    opened_at: Mapped[datetime] = mapped_column(DateTime, default=lambda: datetime.now(timezone.utc), server_default=func.now())

class PaperOrder(Base):
    __tablename__ = "paper_orders_archived"
    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    user_id: Mapped[int] = mapped_column(Integer, nullable=False, index=True)
    symbol: Mapped[str] = mapped_column(String, nullable=False)
    side: Mapped[str] = mapped_column(String, nullable=False)  # buy | sell
    qty: Mapped[float] = mapped_column(Float, nullable=False)
    notional: Mapped[Optional[float]] = mapped_column(Float, nullable=True)  # dollar-based order
    order_type: Mapped[str] = mapped_column(String, nullable=False, default="market")
    limit_price: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    stop_price: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    take_profit_price: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    stop_loss_price: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    trailing_amount: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    trailing_type: Mapped[Optional[str]] = mapped_column(String, nullable=True)  # amount | percent
    tif: Mapped[str] = mapped_column(String, nullable=False, default="day")  # day | gtc | ioc | fok
    extended_hours: Mapped[bool] = mapped_column(Boolean, default=False)
    status: Mapped[str] = mapped_column(String, nullable=False, default="working")  # working | filled | cancelled | rejected | expired
    fill_price: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    fill_qty: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    slippage: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    reject_reason: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    parent_order_id: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)  # for bracket/OCO legs
    filled_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    cancelled_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=lambda: datetime.now(timezone.utc), server_default=func.now())

class PaperTransaction(Base):
    """Realized P&L ledger — one row per fill that closes or reduces a position."""
    __tablename__ = "paper_transactions_archived"
    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    user_id: Mapped[int] = mapped_column(Integer, nullable=False, index=True)
    order_id: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    symbol: Mapped[str] = mapped_column(String, nullable=False)
    side: Mapped[str] = mapped_column(String, nullable=False)  # buy | sell | reset
    qty: Mapped[float] = mapped_column(Float, nullable=False)
    fill_price: Mapped[float] = mapped_column(Float, nullable=False)
    cost_basis: Mapped[float] = mapped_column(Float, nullable=False)  # avg_cost at time of sell
    realized_pnl: Mapped[float] = mapped_column(Float, nullable=False, default=0.0)
    notes: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=lambda: datetime.now(timezone.utc), server_default=func.now())

class PaperDailySnapshot(Base):
    """End-of-day equity snapshot for performance chart."""
    __tablename__ = "paper_daily_snapshots_archived"
    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    user_id: Mapped[int] = mapped_column(Integer, nullable=False, index=True)
    date: Mapped[str] = mapped_column(String, nullable=False)  # YYYY-MM-DD
    equity: Mapped[float] = mapped_column(Float, nullable=False)
    cash: Mapped[float] = mapped_column(Float, nullable=False)
    positions_value: Mapped[float] = mapped_column(Float, nullable=False)
    day_pnl: Mapped[float] = mapped_column(Float, nullable=False, default=0.0)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=lambda: datetime.now(timezone.utc), server_default=func.now())
